### PR TITLE
Separate authors and copyright holders

### PR DESCRIPTION
**Pull request by porsche-rbieniek**
In German law, the author and the copyright holder can be two seperate
legal entities and therefore also need to be treated separately.

Introduce a new copyright holder field that is now the primary source
for copyright holder information. Authors are still only used as
copyright holders if the `addAuthorsToCopyrights` option is enabled.

For now, all package manager implementations set empty copyright
holders. Filling the copyright holder field is left as an exercise for
future actions. Right now, the only way to add copyright holders is via
curations.

This change resolves https://github.com/oss-review-toolkit/ort/issues/4519.

Signed-off-by: Rainer Bieniek <extern.rainer.bieniek@porsche.de>

### CODE REVIEW

**Comment by sschuberth** [comment]
@porsche-rbieniek please have a look at the [detekt issues](https://github.com/oss-review-toolkit/ort/runs/7106829882?check_suite_focus=true#step:4:116).

**Comment by sschuberth** [comment]
@porsche-rbieniek now the DCO checks fails because of a mismatch between the author and sign-off names:

image

**Comment by porsche-rbieniek** [comment]
So many small details to get it right. I was really wondering why my sign-off's were failing

**Comment by sschuberth** [comment]
@porsche-rbieniek Something's got stuck with the GitHub actions apparently... please rebase once more onto origin/main and force-push to retrigger the checks.

**Comment by sschuberth** [comment]
... and please add "Resolves #4519." to the commit message, too.

File: analyzer/src/test/kotlin/managers/CarthageTest.kt
Comment (sschuberth) [inline analyzer/src/test/kotlin/managers/CarthageTest.kt:54]: Please drop the changes to this file. The `authors` aren't asserted here either. I guess felt urged to add these to increase code coverage, this is not required as discussed.

File: analyzer/src/main/kotlin/managers/GoMod.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt:278]: Also here simply: `Go mod doesn't support copyright holders.`

File: model/src/test/kotlin/licenses/LicenseInfoResolverTest.kt
Comment (sschuberth) [inline model/src/test/kotlin/licenses/LicenseInfoResolverTest.kt:689]: The argument should be called `copyrightHolders` also here for consistency with the called function.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/test/kotlin/managers/CarthageTest.kt
Comment (porsche-rbieniek) [reply analyzer/src/test/kotlin/managers/CarthageTest.kt:54]: I would then prefer to assert the authors as well. Enhances the test coverage.

**Comment by porsche-rbieniek** [comment]
Changes implemented except CarthageTest. Here I would like to keep the test cverage and augment it by testing the authors as well.

File: analyzer/src/main/kotlin/managers/GoMod.kt
Comment (porsche-rbieniek) [reply analyzer/src/main/kotlin/managers/GoMod.kt]: Done

File: model/src/test/kotlin/licenses/TestData.kt
Comment (porsche-rbieniek) [reply model/src/test/kotlin/licenses/TestData.kt]: Done

**Comment by porsche-rbieniek** [comment]
Hi guys,

could you please gimme a hint why this seems to be stuck now?

IMHO I implemented the changes requested to move forward

File: analyzer/src/main/kotlin/managers/Bower.kt
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/Bower.kt:256]: An empty set is the default value. Why have you decided to explicitly set it? I've seen that for other properties, like `concludedLicense`, default values are not explicitly provided.

File: model/src/main/kotlin/PackageCurationData.kt
Comment (fviernau) [inline model/src/main/kotlin/PackageCurationData.kt:186]: I believe this change is ok to do in this commit, as the commit is about splitting out `copyrightHolders` from `authors` while in generate I believe that it is wrong: the semantics of the curation is to overwrite any existing copyright holders with the given one. This function deviates from the overwrite semantics. 

Do you agree and if so, shall we add a FIXME comment to fix the issue for both, `authors` and `copyrightHolders`?

File: analyzer/src/main/kotlin/managers/Bower.kt
Comment (porsche-rbieniek) [reply analyzer/src/main/kotlin/managers/Bower.kt:256]: I am aware of that the current default value is to set an empty set. But I am a fan of expressiveness  and I have run a few to oten into the situation that someof changed such a default value in good faith because he "just wanted to solve his problem".

I haven't figured out if there is a hard-an-fast rule in the ORT code base because I observed other places where such a default value was explicitly set.

Bottom line: Is there a real technical or business need to fall back to having this value set though the default?

File: model/src/main/kotlin/PackageCurationData.kt
Comment (porsche-rbieniek) [reply model/src/main/kotlin/PackageCurationData.kt:186]: The intention is to override any copyright finding or authorship  with the values from curation.
The code, as it stands, is merging both authors and copyright findings which contradicts the intened behavior.

IMHO you are actually raising a very good point here that validates a FIXME comment.

File: model/src/main/kotlin/licenses/LicenseInfo.kt
Comment (porsche-rbieniek) [reply model/src/main/kotlin/licenses/LicenseInfo.kt:87]: This 'transferred' actually refers to a legal techniality in Germany.

While by default an author is also the copyright holder, it is possible to 'transfer' the copyright ownership from the author to another legal entity which results in the situation that a author loses any right to copy his own work.

File: scanner/src/main/kotlin/Scanner.kt
Comment (porsche-rbieniek) [reply scanner/src/main/kotlin/Scanner.kt:52]: We should leave a FIXME comment here.

The functionaly will  not get any less broken with this change.

File: model/src/main/kotlin/licenses/LicenseInfo.kt
Comment (fviernau) [reply model/src/main/kotlin/licenses/LicenseInfo.kt:87]: While trying to re-phrase this second sentence, I finally end up voting for dropping it:

The facts that (1) there is means for declaring authors separately from copyrights and (2) the semantics differ 
IMO are good enough arguments for having separate properties. 

IIUC the second sentance was introduced with the goal to explain why there are separate properties for author / copyright holders.

Would you agree dropping makes sense?

**Comment by fviernau** [comment]
@porsche-rbieniek will you make the changes according to the comments so that we can move it forward?

File: scanner/src/main/kotlin/Scanner.kt
Comment (porsche-rbieniek) [reply scanner/src/main/kotlin/Scanner.kt:56]: done

File: model/src/main/kotlin/licenses/LicenseInfoResolver.kt
Comment (porsche-rbieniek) [reply model/src/main/kotlin/licenses/LicenseInfoResolver.kt]: done

File: model/src/main/kotlin/licenses/LicenseInfo.kt
Comment (porsche-rbieniek) [reply model/src/main/kotlin/licenses/LicenseInfo.kt]: Done

File: model/src/main/kotlin/PackageCurationData.kt
Comment (porsche-rbieniek) [reply model/src/main/kotlin/PackageCurationData.kt:188]: Done

**Comment by porsche-rbieniek** [comment]
Done all requested changes, we can proceed

**Review by fviernau — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/Scanner.kt
Comment (fviernau) [reply scanner/src/main/kotlin/Scanner.kt:52]: A dot at the end of the sentence needs to be added.

Comment (fviernau) [reply scanner/src/main/kotlin/Scanner.kt:52]: "The [addAuthorsToCopyrights] configuration property should be taken into account here" for my taste is too much in the solution space. I propose to instead state the issue: "The below condition assumes that `addAuthorsToCopyrights` is enabled, which may not hold."

Rationale: The scanner doesn't even know about `addAuthorsToCopyrights` so maybe the problem has to be solved in a different way than stated currently.

Comment (porsche-rbieniek) [reply scanner/src/main/kotlin/Scanner.kt:52]: I change the wording to 'The below condition partly depends on that addAuthorsToCopyrights is enabled, which may not hold.'

Rationale: The authors are added only if the property is set. But setting copyright owner through curation data is always enabled and independent of the other mechanism.

File: model/src/main/kotlin/PackageCurationData.kt
Comment (porsche-rbieniek) [reply model/src/main/kotlin/PackageCurationData.kt:62]: Also removed the second sentence in the other place I put it before.

**Review by fviernau — APPROVED** [review]

File: model/src/main/kotlin/PackageCurationData.kt
Comment (sschuberth) [inline model/src/main/kotlin/PackageCurationData.kt:179]: "TODO: Clarify semantics about whether setting an author or copyright holders should replace or amend results from a scan."

**Comment by fviernau** [comment]
> While I couldn't join today's developer meeting, I believe we shouldn't try to do too many things at once. IMO the intermediate step, as done in this PR, should be to introduce `copyrightHolders` analog to `authors`, except that `copyrightHolders` for now can exclusively set by curations. All other changes could be done incrementally on top of that; it has already taken us too much time to get this PR into its current shape.

As far as I understand this statement tries to argue against the changes agreed on in the developer meeting.
Maybe it's worth mentioning these here, so that we are sure to be on the same page. Do you know which these are @sschuberth ?

**Comment by sschuberth** [comment]
> Do you know which these are @sschuberth ?

I didn't yet have time to read through the [meeting minutes](https://github.com/oss-review-toolkit/ort/wiki/Developer-Meeting#2022-07-14), if that's what you mean. It seems like use-cases I've never heard of before popped up, and we're back to square one 😵

**Comment by porsche-rishisaxena** [comment]
Hi @fviernau

I wanted to clarify on the following:

1. There was no agreement in the developers meeting. There was an expectation set from the reviewers in this case i.e., you and Thomas who proposed by enhancing the implementation with following cases:
1.1 Implement an ability to add and remove the copyright (c)
1.2 Implement separation of copyright (c) with Authors in the same way as license is being handled
1.3 If the reporter provides copyright (c) for example: "a","b","c" and manually "d" has been identified by the analyst, the curation should only consist of "d" and not "a","b","c", and "d"

2. Thomas requested to change the property name from copyright_holders to declared_copyright_holders and he proposed to change it by himself before merging.

3. @mnonnenmacher even proposed to take this pull request to be merged and later we bring the enhancements after creating an issue in the backlog.

4. Next Steps: we all aligned that we need a separate workshop to brain-storm to have Sebastian, Thomas, Rainer, Frank, Martin and myself to discuss and conclude. Thomas to set up this meeting.

I hope this clarifies.

Regards,
Rishi Saxena

CC: @porsche-rbieniek @sschuberth @tsteenbe @nikpete

**Comment by fviernau** [comment]
> Hi @fviernau
> 
> I wanted to clarify on the following:
> 
>     1. There was no agreement in the developers meeting. There was an expectation set from the reviewers in this case i.e., you and Thomas who proposed by enhancing the implementation with following cases:
>        1.1 Implement an ability to add and remove the copyright (c)
>        1.2 Implement separation of copyright (c) with Authors in the same way as license is being handled
>        1.3 If the reporter provides copyright (c) for example: "a","b","c" and manually "d" has been identified by the analyst, the curation should only consist of "d" and not "a","b","c", and "d"
> 
>     2. Thomas requested to change the property name from copyright_holders to declared_copyright_holders and he proposed to change it by himself before merging.
> 
>     3. @mnonnenmacher even proposed to take this pull request to be merged and later we bring the enhancements after creating an issue in the backlog.
> 
>     4. Next Steps: we all aligned that we need a separate workshop to brain-storm to have Sebastian, Thomas, Rainer, Frank, Martin and myself to discuss and conclude. Thomas to set up this meeting.
> 
> 
> I hope this clarifies.
> 
> Regards, Rishi Saxena
> 
> CC: @porsche-rbieniek @sschuberth @tsteenbe @nikpete

@porsche-rishisaxena 

Hm, so it seems things have been perceived differently, e.g. in your view "There was no agreement in the developers meeting" . In my view, e.g. @mnonnenmacher requested to do #2 as part of the PR, which would contradict your point #2 and #3. 
To my understanding #1 and #4 was discussed as future improvements (and out of scope of the PR), e.g. we've discussed on a conceptual level how we envision the configuration to work in the future.

Anyhow, what problem are you trying to solve @porsche-rishisaxena? Find an agreement how to move on with this PR?

**Comment by porsche-rbieniek** [comment]
Could somebody please explain if there are now further changes required by me or not? 

I couildn't join the developer meeting due to a vacation on short notice and I am now totally lost in this discussion

**Comment by sschuberth** [comment]
> Could somebody please explain if there are now further changes required by me or not?

@porsche-rbieniek, let me take over this PR of yours, I'll get it merged. So please don't push to the branch anymore.

**Comment by tsteenbe** [comment]
@sschuberth  Please don't merge it without my review, I was planning to update this PR today to be alugn with was was discussed in the ORT meeting.

**Comment by mnonnenmacher** [comment]
@sschuberth and @tsteenbe, since your are taking over the PR, to me the most important point is consistency in the handling of copyrights and licenses. Both can be declared in the metadata and detected in the source code, so the properties should use similar naming and the curation options should work in a similar way.
* Similar to `declaredLicense` I would prefer the name `declaredCopyrights` over `copyrightHolders` in `Package` and `Project`.
* I prefer `copyrights` over `copyrightHolders` because the detected values are statements like "(C) 2020 xyz", but "holder" sounds to me like it should only be "xyz" for this example. And likely the values in the metadata are also similar in structure as the values we get from ScanCode. See for example "© Microsoft Corporation.  All rights reserved." here: https://api.nuget.org/v3/catalog0/data/2018.12.13.22.08.00/system.globalization.4.3.0.json (Btw, who will actually do the work and go through all package managers and check which have dedicated fields for copyright information?)
* I don't know if it's currently possible to add declared licenses with a curation, but if this is implemented for copyrights it should work for licenses and copyrights in a similar way.

**Comment by tsteenbe** [comment]
@mnonnenmacher  I looked up if the most used package manager have a declared copyright holder filed and they don't.

Cargo <https://doc.rust-lang.org/cargo/reference/manifest.html>
- authors — The authors of the package.
- license — The package license.
- license-file — Path to the text of the license.
Composer: <https://getcomposer.org/doc/04-schema.md>
- license
- authors 
Maven: <https://maven.apache.org/pom.html#More_Project_Information>
- license
- organization
- developer
- contributor
NPM:
- license: SPDX license id <https://docs.npmjs.com/cli/v8/configuring-npm/package-json#license>
- author: one person <https://docs.npmjs.com/cli/v8/configuring-npm/package-json#people-fields-author-contributors>
- contributor: array of people
Python: <https://docs.python.org/3/distutils/setupscript.html#additional-meta-data>
- author
- author_email
- maintainer
- maintainer_email
- license

**Comment by sschuberth** [comment]
Superseded by https://github.com/oss-review-toolkit/ort/pull/5680.

**Pull request closed without merging**
